krm-functions/nfdeploy-fn/common/helper.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2023 The Nephio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	nephiodeployv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+	nephioreqv1alpha1 "github.com/nephio-project/api/nf_requirements/v1alpha1"
+	kptcondsdk "github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sort"
+)
+
+type NfDeployFn[T NfType] struct {
+	sdk                 kptcondsdk.KptCondSDK
+	site                string
+	gvk                 schema.GroupVersionKind
+	capacity            nephioreqv1alpha1.Capacity
+	pkgName             string
+	networkInstance     map[string]nephiodeployv1alpha1.NetworkInstance
+	interfaceConfigsMap map[string]nephiodeployv1alpha1.InterfaceConfig
+}
+
+func NewMutator[T NfType](gvk schema.GroupVersionKind) NfDeployFn[T] {
+	return NfDeployFn[T]{
+		interfaceConfigsMap: make(map[string]nephiodeployv1alpha1.InterfaceConfig),
+		networkInstance:     make(map[string]nephiodeployv1alpha1.NetworkInstance),
+		gvk:                 gvk,
+	}
+}
+
+func (h *NfDeployFn[T]) SetInterfaceConfig(interfaceConfig nephiodeployv1alpha1.InterfaceConfig, networkInstanceName string) {
+	// dont add to empty networkInstanceName, ideally should not happen
+	if networkInstanceName == "" {
+		return
+	}
+
+	h.interfaceConfigsMap[networkInstanceName] = interfaceConfig
+}
+
+func (h *NfDeployFn[T]) AddDNNToNetworkInstance(dnn nephiodeployv1alpha1.DataNetwork, networkInstanceName string) {
+	if nI, ok := h.networkInstance[networkInstanceName]; ok {
+		nI.DataNetworks = append(nI.DataNetworks, dnn)
+		h.networkInstance[networkInstanceName] = nI
+	} else {
+		h.networkInstance[networkInstanceName] = nephiodeployv1alpha1.NetworkInstance{
+			Name:         networkInstanceName,
+			DataNetworks: []nephiodeployv1alpha1.DataNetwork{dnn},
+		}
+	}
+}
+
+func (h *NfDeployFn[T]) AddInterfaceToNetworkInstance(interfaceName, networkInstanceName string) {
+	if nI, ok := h.networkInstance[networkInstanceName]; ok {
```

Comment:
Direct assignment is not possible since its NetworkInstance is not pointer. 

Ref:
https://stackoverflow.com/questions/32751537/why-do-i-get-a-cannot-assign-error-when-setting-value-to-a-struct-as-a-value-i